lib/Dialect/Secret/Conversions/SecretToCKKS/SecretToCKKS.cpp

```diff
@@ -269,6 +272,105 @@ class SecretGenericTensorInsertConversion
+ public:
+  using OpConversionPattern<affine::AffineForOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      affine::AffineForOp forOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    const auto *contextAwareTypeConverter =
+        dynamic_cast<const TypeWithAttrTypeConverter *>(getTypeConverter());
+    SmallVector<Type> newInitTypes;
+    contextAwareTypeConverter->convertValueRangeTypes(forOp.getInits(),
+                                                      newInitTypes);
+
+    Location loc = forOp.getLoc();
+    auto newForOp = rewriter.create<affine::AffineForOp>(
+        loc, ValueRange(forOp.getLowerBoundOperands()),
+        forOp.getLowerBoundMap(), ValueRange(forOp.getUpperBoundOperands()),
+        forOp.getUpperBoundMap(), forOp.getStep().getZExtValue(),
+        adaptor.getInits(), [](OpBuilder &, Location, Value, ValueRange) {});
+    newForOp->setAttrs(forOp->getAttrs());
+
+    mlir::Block *newBody = newForOp.getBody();
+    mlir::Block *oldBody = forOp.getBody();
+    rewriter.setInsertionPoint(newBody, newBody->begin());
+
+    SmallVector<Value> newBlockArgs;
+    for (auto arg : newBody->getArguments()) {
+      if (auto lweTy = dyn_cast<lwe::NewLWECiphertextType>(arg.getType())) {
+        // For each block arg that is a secret type, we need to create an
+        // operation inside the for loops block that can hold the mgmt attr for
+        // type conversion. We can't rely on getValueAttr since blocks are
+        // unlinked during dialect conversion.
+        auto underlyingTy = secret::SecretType::get(
+            lweTy.getApplicationData().getMessageType());
+        auto cast = rewriter.create<mlir::UnrealizedConversionCastOp>(
+            loc, underlyingTy, arg);
+        cast->setAttrs(oldBody->getArgument(arg.getArgNumber())
+                           .getUsers()
+                           .begin()
+                           ->getAttrs());
+        newBlockArgs.push_back(cast.getResult(0));
+      } else {
+        newBlockArgs.push_back(arg);
+      }
+    }
+
+    // Move the body of the old ForOp to the new one.
+    rewriter.mergeBlocks(oldBody, newBody, newBlockArgs);
+    rewriter.replaceOp(forOp, newForOp);
+    return success();
+  }
+};
+
+class SecretGenericFuncCallConversion
+    : public SecretGenericOpConversion<func::CallOp, func::CallOp> {
+ public:
+  using SecretGenericOpConversion<func::CallOp,
+                                  func::CallOp>::SecretGenericOpConversion;
+
+  LogicalResult matchAndRewriteInner(
+      secret::GenericOp op, TypeRange outputTypes, ValueRange inputs,
+      ArrayRef<NamedAttribute> attributes,
+      ConversionPatternRewriter &rewriter) const override {
+    // check if any args are secret from wrapping generic
+    // clone the callee (and update a unique name, for now always) the call
+    // operands add a note that we don't have to always clone to be secret
+    // update the called func's type signature
```

Comment:
This is mildly surprising. I would have expected a pattern that hooks on `func.func` to be responsible for updating its own signature.

Would it be safer, but still work for your use case, to restrict the signature conversion to only apply to funcs that are declarations? Or maybe mark cleaning this up as TODO for https://github.com/google/heir/issues/1311, where we will have a more comprehensive support of programs that have multiple functions

Reply:
I missed this! Creating a second PR for this - but I see what you mean, I could also hook on func.func and check for callers in secret.generics. That tweaks how legality is checked (to instead check if downstream users of the func.func are calling it with secret args).

> Would it be safer, but still work for your use case, to restrict the signature conversion to only apply to funcs that are declarations? 

That's a nice restriction, since I see that this trick doesn't work with functions that have a region.